xtask/src/codegen/gen_diagnostic_docs.rs

```diff
@@ -0,0 +1,73 @@
+//! Generates `assists.md` documentation.
+
+use std::{fmt, fs, path::PathBuf};
+
+use crate::{
+    codegen::{self, extract_comment_blocks_with_empty_lines, Location, Mode, PREAMBLE},
+    project_root, rust_files, Result,
+};
+
+pub fn generate_diagnostic_docs(mode: Mode) -> Result<()> {
+    let features = Diagnostic::collect()?;
```

Comment:
Copy-paste detected :-)

This is fine for such code, though pulling common "traverse the files and colllect magic comments" might be pulled into some helper function somewhere.

Reply:
😊

Reply:
Regarding the pulling common code: I think it'll be better to not mix refactoring and new functionality in one PR, if that's possible.